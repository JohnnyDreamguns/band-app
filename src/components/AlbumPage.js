import React from 'react';
import { Link } from 'react-router-dom';

const AlbumPage = ({ loading, album, band }) => (
  <div>
    {loading && <div className="loader"></div>}

    {album && band && (
      <div>
        <h2>{album.name}</h2>
        <p>Released: {album.released}</p>

        <p>
          Back to <Link to={`/band/${band.id}`}>{band.name}</Link>
        </p>
      </div>
    )}
  </div>
);

export default AlbumPage;
