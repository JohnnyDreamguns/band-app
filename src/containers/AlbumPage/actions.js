import {
  GET_ALBUM_BY_ID,
  SET_ALBUM_ID,
  CLEAR_ALBUM_PAGE,
  SET_LOADING
} from './constants';

export const getAlbumById = payload => ({
  type: GET_ALBUM_BY_ID,
  payload
});

export const setAlbumId = payload => ({
  type: SET_ALBUM_ID,
  payload
});

export const clearAlbumpage = () => ({
  type: CLEAR_ALBUM_PAGE
});

export const setLoading = payload => ({
  type: SET_LOADING,
  payload
});
