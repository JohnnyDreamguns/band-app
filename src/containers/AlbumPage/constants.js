export const GET_ALBUM_BY_ID = 'bands-app/AlbumPage/GET_ALBUM_BY_ID';
export const GET_BAND_BY_ID = 'bands-app/AlbumPage/GET_BAND_BY_ID';
export const SET_ALBUM_ID = 'bands-app/AlbumPage/SET_ALBUM_ID';
export const CLEAR_ALBUM_PAGE = 'bands-app/AlbumPage/CLEAR_ALBUM_PAGE';
export const SET_LOADING = 'bands-app/AlbumPage/SET_LOADING';
