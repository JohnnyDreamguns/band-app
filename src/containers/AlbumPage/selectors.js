import { initialState } from './reducer';
import { selectBandsData, selectAlbumsData } from '../../selectors/data';

export const albumRoot = state => state.albumPage || initialState;

export const selectAlbumId = state => albumRoot(state)['albumId'];

export const selectAlbum = state => {
  const albumId = selectAlbumId(state);
  if (!albumId) return undefined;
  return selectAlbumsData(state)[`${albumId}`];
};

export const selectLoading = state => albumRoot(state)['loading'];

export const selectBandId = state => {
  const album = selectAlbum(state);
  if (!album) return undefined;
  return album['band'];
};

export const selectBand = state => {
  const bandId = selectBandId(state);
  if (!bandId) return undefined;
  return selectBandsData(state)[`${bandId}`];
};
