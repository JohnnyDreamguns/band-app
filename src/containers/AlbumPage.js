import React, { useEffect } from 'react';
import AlbumPage from '../components/AlbumPage';
import useAlbumData from '../hooks/useAlbumData';

const AlbumPageContainer = props => {
  const {
    album,
    band,
    loading,
    fetchAlbumPageData,
    clearAlbumId
  } = useAlbumData();
  const { albumId } = props.match.params;

  useEffect(() => {
    fetchAlbumPageData(albumId);
    return () => {
      clearAlbumId();
    };
  }, [albumId, fetchAlbumPageData, clearAlbumId]);

  const componentProps = {
    album,
    band,
    loading
  };

  return <AlbumPage {...componentProps} />;
};

export default AlbumPageContainer;
