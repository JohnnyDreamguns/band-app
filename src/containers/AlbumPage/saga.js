import { call, put, takeLatest, select } from 'redux-saga/effects';
import { GET_ALBUM_BY_ID } from './constants';
import { setLoading, setAlbumId } from './actions';
import { setError } from '../../actions/messages';
import { setBands, setAlbums } from '../../actions/data';
import { selectBandsData } from '../../selectors/data';

import request from '../../utils/request';

export function* getAlbumById(action) {
  const albumsURL = `http://localhost:3001/albums`;

  try {
    yield put(setLoading(true));

    const albumId = action.payload;

    const album = yield call(request, `${albumsURL}/${albumId}`);
    yield put(setAlbums(album.results));
    yield put(setAlbumId(parseInt(Object.keys(album.results)[0])));

    const bandId = album.results[albumId].band;
    const bandsData = yield select(selectBandsData);

    if (!bandsData.hasOwnProperty(bandId)) {
      yield getBandById(bandId);
    }

    yield put(setLoading(false));
  } catch (err) {
    yield put(setLoading(false));
    yield put(setError('There was a problem loading the album data'));
  }
}

export function* getBandById(bandId) {
  const bandsURL = `http://localhost:3001/bands`;

  const band = yield call(request, `${bandsURL}/${bandId}`);
  yield put(setBands(band.results));
}

export default function* bandpageDataSaga() {
  yield takeLatest(GET_ALBUM_BY_ID, getAlbumById);
}
