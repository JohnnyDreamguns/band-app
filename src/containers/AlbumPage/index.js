import React, { useEffect } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { getAlbumById, clearAlbumpage } from './actions';
import { useInjectSaga } from '../../utils/injectSaga';
import { useInjectReducer } from '../../utils/injectReducer';
import { selectLoading, selectAlbum, selectBand } from './selectors';
import saga from './saga';
import reducer from './reducer';

const key = 'albumPage';

export const AlbumPage = ({
  albumId,
  loading,
  album,
  band,
  getAlbumById,
  clearAlbumpage
}) => {
  useInjectReducer({ key, reducer });
  useInjectSaga({ key, saga });

  useEffect(() => {
    getAlbumById(albumId);

    return () => {
      clearAlbumpage();
    };
  }, [getAlbumById, clearAlbumpage, albumId]);

  return (
    <div>
      {loading && <div className="loader"></div>}

      {album && band && (
        <div>
          <h2>{album.name}</h2>
          <p>Released: {album.released}</p>

          <p>
            Back to <Link to={`/band/${band.id}`}>{band.name}</Link>
          </p>
        </div>
      )}
    </div>
  );
};

export const mapStateToProps = (state, ownProps) => {
  const { albumId } = ownProps.match.params;

  return {
    albumId,
    loading: selectLoading(state),
    album: selectAlbum(state),
    band: selectBand(state)
  };
};

export const mapDispatchToProps = dispatch => ({
  getAlbumById: albumId => dispatch(getAlbumById(albumId)),
  clearAlbumpage: () => dispatch(clearAlbumpage())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AlbumPage);
