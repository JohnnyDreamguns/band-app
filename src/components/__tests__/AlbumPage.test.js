import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import AlbumPage from '../AlbumPage';

const renderer = new ShallowRenderer();

const props = {
  albumId: 2,
  loading: false,
  album: {
    name: 'Album Name',
    released: '2009'
  },
  band: {
    id: 1,
    name: 'Band Name'
  }
};

describe('<AlbumPage />', () => {
  it('should match the snapshot', () => {
    renderer.render(<AlbumPage {...props} />);
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });

  it('should show loading indicator', () => {
    renderer.render(<AlbumPage {...props} loading={true} />);
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });

  it('should show no content if band data is missing', () => {
    renderer.render(<AlbumPage {...props} band={undefined} />);
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });

  it('should show no content if album data is missing', () => {
    renderer.render(<AlbumPage {...props} album={undefined} />);
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });
});
