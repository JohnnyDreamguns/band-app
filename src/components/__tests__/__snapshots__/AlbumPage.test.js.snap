// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AlbumPage /> should match the snapshot 1`] = `
<div>
  <div>
    <h2>
      Album Name
    </h2>
    <p>
      Released: 
      2009
    </p>
    <p>
      Back to 
      <ForwardRef
        to="/band/1"
      >
        Band Name
      </ForwardRef>
    </p>
  </div>
</div>
`;

exports[`<AlbumPage /> should show loading indicator 1`] = `
<div>
  <div
    className="loader"
  />
  <div>
    <h2>
      Album Name
    </h2>
    <p>
      Released: 
      2009
    </p>
    <p>
      Back to 
      <ForwardRef
        to="/band/1"
      >
        Band Name
      </ForwardRef>
    </p>
  </div>
</div>
`;

exports[`<AlbumPage /> should show no content if album data is missing 1`] = `<div />`;

exports[`<AlbumPage /> should show no content if band data is missing 1`] = `<div />`;
